include/pybind11/attr.h

```diff
@@ -14,27 +14,22 @@
 
 NAMESPACE_BEGIN(pybind11)
 
-template <typename T> struct arg_t;
-
 /// Annotation for keyword arguments
 struct arg {
     constexpr explicit arg(const char *name) : name(name) { }
 
     template <typename T>
     constexpr arg_t<T> operator=(const T &value) const { return {name, value}; }
-    template <typename T, size_t N>
-    constexpr arg_t<const T *> operator=(T const (&value)[N]) const {
-        return operator=((const T *) value);
-    }
 
     const char *name;
 };
 
 /// Annotation for keyword arguments with default values
 template <typename T> struct arg_t : public arg {
     constexpr arg_t(const char *name, const T &value, const char *descr = nullptr)
-        : arg(name), value(value), descr(descr) { }
-    T value;
+        : arg(name), value(&value), descr(descr) { }
```

Comment:
Temporary objects (even ones which are not bound to a `const&`) are only destroyed after the full expression is evaluated. A full expression is one which is not contained within any other expression (so essentially, temp object lifetime is extended to a semicolon).

This makes the following line perfectly safe:
```
f("arg"_a=MyArg(1, 2, 3));
```
On the other hand,
```
auto a = "arg"_a=MyArg(1, 2, 3);
f(a);
```
would not be safe, but this is currently a compile time error because `unpacking_collector` only accepts rvalues: `arg<T>&&`. (Although, I do need to add this safeguard to `process_attribute` as well.)

That said, I think the ideal implementation of `arg_t` would be to hold a `py::object` instead of `T`:
```c++
struct arg_t : arg {
    template <typename T>
    arg_t(const char *name, const T &value, const char *descr = nullptr)
        : arg(name), value(make_caster<T>::cast(value), return_value_policy::automatic), descr(descr) { }
    object value;
    const char *descr;
}
```

The to-python cast would then be done right away instead of being delayed. I didn't make this change originally because default arguments are collected with the `automatic` policy while function calls use `automatic_reference` by default. Would there be any downside to always using `automatic` for call argument collection? (Users should be able to override by doing a manual `py::cast(value, policy)` when assigning an argument.)